File: pos/src/lib/weekend-promo.test.ts
import { describe, expect, it } from "vitest";
import {
  activePromoInfo,
  eligiblePromoSubtotal,
  isWeekendPromoDay,
  setDiscountRulesCache,
  weekendDiscount,
  weekendPromoLabel,
} from "@/lib/weekend-promo";

describe("weekend promo (Fri & Sun 10% non-deal until 31 Aug 2026)", () => {
  it("applies only Fri/Sun Asia/Karachi (not Saturday)", () => {
    setDiscountRulesCache([]);
    expect(isWeekendPromoDay(new Date("2026-08-07T12:00:00+05:00"))).toBe(
      true,
    );
    expect(isWeekendPromoDay(new Date("2026-08-08T12:00:00+05:00"))).toBe(
      false,
    );
    expect(isWeekendPromoDay(new Date("2026-08-09T12:00:00+05:00"))).toBe(
      true,
    );
    expect(isWeekendPromoDay(new Date("2026-08-03T12:00:00+05:00"))).toBe(
      false,
    );
  });

  it("turns off after 31 Aug 2026 even on Friday/Sunday", () => {
    setDiscountRulesCache([]);
    expect(isWeekendPromoDay(new Date("2026-09-04T12:00:00+05:00"))).toBe(
      false,
    );
    expect(isWeekendPromoDay(new Date("2026-08-31T12:00:00+05:00"))).toBe(
      false,
    );
    expect(isWeekendPromoDay(new Date("2026-08-30T18:00:00+05:00"))).toBe(
      true,
    );
  });

  it("excludes deal items from eligible subtotal", () => {
    const eligible = eligiblePromoSubtotal([
      { product_name: "Chicken Tikka", price: 800, quantity: 1, is_deal: false },
      { product_name: "Family Deal", price: 2000, quantity: 1, is_deal: true },
      { product_name: "Mega Combo Box", price: 1500, quantity: 1 },
    ]);
    expect(eligible).toBe(800);
  });

  it("gives 10% when eligible subtotal >= 1000 on Fri/Sun", () => {
    setDiscountRulesCache([]);
    const friday = new Date("2026-08-07T18:00:00+05:00");
    const discount = weekendDiscount(
      [
        { product_name: "Pizza", price: 600, quantity: 2, is_deal: false },
        {
          product_name: "Deal Special",
          price: 1999,
          quantity: 1,
          is_deal: true,
        },
      ],
      friday,
    );
    expect(discount).toBe(120);
    expect(weekendPromoLabel(undefined, friday)).toMatch(/Fri & Sun 10%/);
  });

  it("gives zero on Saturday even above min", () => {
    setDiscountRulesCache([]);
    const saturday = new Date("2026-08-08T18:00:00+05:00");
    expect(
      weekendDiscount(
        [{ product_name: "Pizza", price: 2000, quantity: 1 }],
        saturday,
      ),
    ).toBe(0);
  });

  it("gives zero below Rs 1000 eligible or on weekday", () => {
    setDiscountRulesCache([]);
    const friday = new Date("2026-08-07T18:00:00+05:00");
    const monday = new Date("2026-08-03T18:00:00+05:00");
    expect(
      weekendDiscount(
        [{ product_name: "Pizza", price: 500, quantity: 1 }],
        friday,
      ),
    ).toBe(0);
    expect(
      weekendDiscount(
        [{ product_name: "Pizza", price: 2000, quantity: 1 }],
        monday,
      ),
    ).toBe(0);
  });

  it("uses cached Azaadi rule name and min for labels", () => {
    const thursday = new Date("2026-08-13T12:00:00+05:00");
    setDiscountRulesCache([
      {
        name: "Azaadi Discount",
        active: true,
        percent: 20,
        min_subtotal: 2000,
        schedule_type: "date_range",
        start_date: "2026-08-13",
        end_date: "2026-08-14",
        exclude_deals: true,
      },
    ]);
    const info = activePromoInfo(undefined, thursday);
    expect(info?.name).toBe("Azaadi Discount");
    expect(info?.min_subtotal).toBe(2000);
    expect(
      weekendDiscount(
        [{ product_name: "Pizza", price: 3330, quantity: 1 }],
        thursday,
      ),
    ).toBe(666);
    expect(
      weekendPromoLabel(
        [{ product_name: "Pizza", price: 3330, quantity: 1 }],
        thursday,
      ),
    ).toBe("Azaadi Discount");
    setDiscountRulesCache([]);
  });
});

File: backend/internal/service/weekend_promo_test.go
package service

import (
	"testing"
	"time"
)

func TestWeekendPromoActive_FriSunOnlyUntilAug31(t *testing.T) {
	loc, err := time.LoadLocation("Asia/Karachi")
	if err != nil {
		loc = time.FixedZone("PKT", 5*3600)
	}
	at := func(y int, m time.Month, d, hh int) time.Time {
		return time.Date(y, m, d, hh, 0, 0, 0, loc)
	}

	if !WeekendPromoActive(at(2026, time.August, 7, 12)) { // Fri
		t.Fatal("Friday Aug 7 should be active")
	}
	if WeekendPromoActive(at(2026, time.August, 8, 12)) { // Sat
		t.Fatal("Saturday should not be active")
	}
	if !WeekendPromoActive(at(2026, time.August, 9, 12)) { // Sun
		t.Fatal("Sunday Aug 9 should be active")
	}
	if !WeekendPromoActive(at(2026, time.August, 30, 20)) { // Sun last weekend of Aug
		t.Fatal("Sunday Aug 30 should still be active")
	}
	if WeekendPromoActive(at(2026, time.September, 4, 12)) { // Fri after end
		t.Fatal("Friday after 31 Aug should be inactive")
	}
	if WeekendPromoActive(at(2026, time.August, 3, 12)) { // Mon
		t.Fatal("Monday should not be active")
	}
}

func TestWeekendDiscount_Amounts(t *testing.T) {
	loc, _ := time.LoadLocation("Asia/Karachi")
	friday := time.Date(2026, time.August, 7, 18, 0, 0, 0, loc)
	saturday := time.Date(2026, time.August, 8, 18, 0, 0, 0, loc)

	if got := WeekendDiscount(friday, 1200); got != 120 {
		t.Fatalf("Fri 1200 → want 120 got %d", got)
	}
	if got := WeekendDiscount(friday, 999); got != 0 {
		t.Fatalf("below min → want 0 got %d", got)
	}
	if got := WeekendDiscount(saturday, 5000); got != 0 {
		t.Fatalf("Sat → want 0 got %d", got)
	}
}
